semgrep-core/src/engine/Match_extract_mode.ml

```diff
@@ -0,0 +1,363 @@
+  let f : Common.dirname = Common.new_temp_file "extracted" dst_lang in
+  Common2.write_file ~file:f contents;
+  { In.path = f; language = dst_lang; rule_ids }
+
+(*****************************************************************************)
+(* Error reporting *)
+(*****************************************************************************)
+
+let report_unbound_mvar ruleid mvar m =
+  let { Range.start; end_ } = Pattern_match.range m in
+  logger#warning
+    "The extract metavariable for rule %s (%s) wasn't bound in a match; \
+     skipping extraction for this match [match was at bytes %d-%d]"
+    ruleid mvar start end_
+
+let report_no_source_range erule =
+  logger#error
+    "In rule %s the extract metavariable (%s) did not have a corresponding \
+     source range"
+    (fst erule.Rule.id)
+    (let (`Extract { Rule.extract; _ }) = erule.mode in
+     extract)
+
+(*****************************************************************************)
+(* Result mapping helpers *)
+(*****************************************************************************)
+
+let map_loc pos line col file (loc : Parse_info.token_location) =
+  (* this _shouldn't_ be a fake location *)
+  {
+    loc with
+    charpos = loc.charpos + pos;
+    line = loc.line + line;
+    column = (if loc.line = 1 then loc.column + col else loc.column);
+    file;
+  }
+
+let map_res map_loc tmpfile file
+    (mr : Report.partial_profiling Report.match_result) =
+  {
+    Report.matches =
+      Common.map
+        (fun (m : Pattern_match.t) ->
+          { m with file; range_loc = Common2.pair map_loc m.range_loc })
+        mr.matches;
+    errors =
+      Common.map
+        (fun (e : Semgrep_error_code.error) -> { e with loc = map_loc e.loc })
+        mr.errors;
+    skipped_targets =
+      Common.map
+        (fun (st : Output_from_core_t.skipped_target) ->
+          { st with path = (if st.path = tmpfile then file else st.path) })
+        mr.skipped_targets;
+    profiling = { mr.profiling with Report.file };
+  }
+
+(*****************************************************************************)
+(* Main logic *)
+(*****************************************************************************)
+
+let extract_collect_and_comb erule_table xtarget rule_ids matches =
+  matches
+  (* Group the matches within this file by rule id *)
+  |> collect (fun m m' -> m.Pattern_match.rule_id = m'.Pattern_match.rule_id)
+  |> Common.map (fun matches -> nonempty_to_list matches)
+  (* Convert matches to the extract metavariable / bound value *)
+  |> Common.map
+       (List.filter_map (fun m ->
+            match extract_of_match erule_table m with
+            | Some ({ mode = `Extract { Rule.extract; _ }; id = id, _; _ }, None)
+              ->
+                report_unbound_mvar id extract m;
+                None
+            | Some (r, Some mval) -> Some (r, mval)
+            | None -> None))
+  (* Factor out rule *)
+  |> List.filter_map (function
+       | [] -> None
```

Comment:
Just a question: why not factor out the rule earlier, when you still have nonempty lists?

Reply:
We don't actually have the rule itself until we look it up in the table in the step immediately prior to this; we just have the ruleid on each match before then.